explainer.md

```diff
@@ -0,0 +1,87 @@
+
+# Problem Statement
+There is currently no way for a site to indicate that it has a custom transition which should override the default UA transition. For example, if the UA adds a visual transition for an atomic navigations (like clicking the back button) which the site has also customized.
+
+## Same-Document Navigations
+For same-document navigations the UA can not detect whether a site will execute a visual transition for a navigation. Also, while the site can not currently customize the swipe gesture, it may still do a visual transition when the navigation is eventually committed. Adding UA transitions for these navigations has a compat risk of "double transitions". This [site](https://darkened-relieved-azimuth.glitch.me) shows the issue on Chrome/Safari iOS and Safari on Mac. The site does a cross-fade when the user goes back:
+
+* If the user navigates using the back button in the UI, a cross-fade defined by the site executes and is the only transition.
+
+* If the user navigates using a swipe from the edge (iOS) or 2 finger swipe (mac), the UA does a transition showing content of the back entry. The Document then receives a `popState` event for the navigation and performs another visual transition.
+
+## Cross-Document Navigations
+This is not a problem for cross-document navigations since customization requires UA primitives. The UA can use this to detect if there are custom transitions and give them precedence. For example:
+
+* Since atomic navigations will require a ViewTransition, the UA transition is suppressed only if the current Document has an opt-in for ViewTransitions for a navigation.
+
+* Since swipe navigations will require a combination of a Gesture API + ViewTransition, the UA transition is suppressed only if the current Document has an opt-in for both. If there is an opt-in for only one of them and a swipe occurs, the opt-in is ignored.
+
+# Proposals
+## Choosing between UA and Custom Transition
+This API provides authors control over whether a same-document navigation uses a UA vs custom visual transition. A global setting called `same-document-ua-transition` is introduced with the following options:
+
+* `disable`: Suppresses UA transitions for all navigations.
+* `enable`: Enables UA transitions for all navigations.
+* `enable-atomic`: Enables UA transitions for atomic navigations only.
+* `enable-swipe`: Enables UA transitions for swipe navigations only.
+* `auto`: Allows the UA to decide which navigations get a UA transition.
+
+`auto` is the default value for this setting. This allows UAs to make different decisions based on their current behaviour. For example, Safari already ships with transitions on swipe gestures; changing that would surprise users.
+
+Question: It seems awkward to have this setting dictate behaviour for swipes. Especially since the Gesture API for swipes will need their own opt-in.
```

Comment:
If a swipe API like this existed, here's how I would think about this:

* The swipe API has its own opt-in, maybe through an event listener, which decides whether you're customizing that case. Its possible that you decide not to take over the swipe when the event is dispatched, in which case its as if you didn't opt-in.
* `same-document-ua-transition` setting is a global setting which tells the UA if you have any transitions at all, with just 2 values: enable and disable.
* If the value is enable, UA defaults always win and you don't get swipe events.
* If the value is disable but you haven't opted-in to the swipe API, then UA transitions win for swipe but nothing else.
* If the value is disable and you have opted-in to the swipe API, then no UA transitions on all navigations.

This is trying to follow a pattern similar to cross-document navigations. If the author has to rely on a primitive to do the transition, we can detect customization based on whether the site is using that primitive.

Reply:
Btw the default value we choose can be up for debate. I suspect we'll need `auto` since differences between browsers means we might not align on what the default should be.